x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,359 @@
+    public static final ParseField PRUNING_CONFIG_FIELD = new ParseField("pruning_config");
+
+    private static final boolean DEFAULT_PRUNE = false;
+    private final String fieldName;
+    private final List<WeightedToken> queryVectors;
+    private final String inferenceId;
+    private final String text;
+    private final Boolean shouldPruneTokens;
+
+    private SetOnce<TextExpansionResults> weightedTokensSupplier;
+
+    @Nullable
+    private final TokenPruningConfig tokenPruningConfig;
+
+    public SparseVectorQueryBuilder(String fieldName, String inferenceId, String text) {
+        this(fieldName, null, inferenceId, text, DEFAULT_PRUNE, null);
+    }
+
+    public SparseVectorQueryBuilder(String fieldName, List<WeightedToken> queryVector) {
+        this(fieldName, queryVector, null, null, DEFAULT_PRUNE, null);
+    }
+
+    public SparseVectorQueryBuilder(
+        String fieldName,
+        @Nullable List<WeightedToken> queryVectors,
+        @Nullable String inferenceId,
+        @Nullable String text,
+        @Nullable Boolean shouldPruneTokens,
+        @Nullable TokenPruningConfig tokenPruningConfig
+    ) {
+        this.fieldName = Objects.requireNonNull(fieldName, "[" + NAME + "] requires a field");
+        this.shouldPruneTokens = (shouldPruneTokens != null ? shouldPruneTokens : DEFAULT_PRUNE);
+        this.queryVectors = queryVectors;
+        this.inferenceId = inferenceId;
+        this.text = text;
+        this.tokenPruningConfig = tokenPruningConfig;
+
+        if (queryVectors == null ^ inferenceId == null == false) {
+            throw new IllegalArgumentException(
+                "["
+                    + NAME
+                    + "] requires one of ["
+                    + QUERY_VECTOR_FIELD.getPreferredName()
+                    + "] or ["
+                    + INFERENCE_ID_FIELD.getPreferredName()
+                    + "]"
+            );
+        }
+        if (inferenceId != null && text == null) {
+            throw new IllegalArgumentException(
+                "["
+                    + NAME
+                    + "] requires ["
+                    + TEXT_FIELD.getPreferredName()
+                    + "] when ["
+                    + INFERENCE_ID_FIELD.getPreferredName()
+                    + "] is specified"
+            );
+        }
+    }
+
+    public SparseVectorQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.shouldPruneTokens = in.readOptionalBoolean();
+        this.queryVectors = in.readOptionalCollectionAsList(WeightedToken::new);
+        this.inferenceId = in.readOptionalString();
+        this.text = in.readOptionalString();
+        this.tokenPruningConfig = in.readOptionalWriteable(TokenPruningConfig::new);
+    }
+
+    public SparseVectorQueryBuilder(SparseVectorQueryBuilder other, SetOnce<TextExpansionResults> weightedTokensSupplier) {
```

Comment:
This should be private.
```suggestion
    private SparseVectorQueryBuilder(SparseVectorQueryBuilder other, SetOnce<TextExpansionResults> weightedTokensSupplier) {
```